Sources/MongoSwift/MongoConnectionString.swift

```diff
@@ -0,0 +1,107 @@
+/// Represents a MongoDB connection string.
+/// - SeeAlso: https://docs.mongodb.com/manual/reference/connection-string/
+public struct MongoConnectionString: Codable, LosslessStringConvertible {
+    /// Represents a connection string scheme.
+    public struct Scheme: LosslessStringConvertible, Equatable {
+        /// Indicates that this connection string uses the scheme `mongodb`.
+        public static let mongodb = Scheme(.mongodb)
+
+        /// Indicates that this connection string uses the scheme `mongodb+srv`.
+        public static let srv = Scheme(.srv)
+
+        /// Internal representation of a scheme.
+        private enum _Scheme: String {
+            case mongodb = "mongodb"
+            case srv = "mongodb+srv"
+        }
+
+        private let _scheme: _Scheme
+
+        private init(_ value: _Scheme) {
+            self._scheme = value
+        }
+
+        /// LosslessStringConvertible` protocol requirements
+        public init?(_ description: String) {
+            guard let _scheme = _Scheme(rawValue: description) else {
+                return nil
+            }
+            self.init(_scheme)
+        }
+
+        public var description: String { _scheme.rawValue }
+    }
+
+    /// Parses `input` into its constituent parts.
+    /// - Throws:
+    ///   - `MongoError.InvalidArgumentError` if the input is invalid.
+    public init(throwsIfInvalid input: String) throws {
+        let getScheme = input.components(separatedBy: "://")
+        guard getScheme.count > 1, let scheme = Scheme(getScheme[0]) else {
+            throw MongoError.InvalidArgumentError(
+                message: "Invalid URI Schema, expecting \'mongodb://\' or \'mongodb+srv://\'"
+            )
+        }
+        let getUser = getScheme[1].components(separatedBy: "@")
+        let userInfo = getUser.count > 1 ?  getUser[0].components(separatedBy: ":") : nil
+        let getHost = getUser.count > 1 ? getUser[1].components(separatedBy: "/") :
```

Comment:
feels like im repeating some code here `getUser.count > 1`

Reply:
for now, let's not worry about userinfo parsing, as that will be handled in the next PR for SWIFT-1161. For this PR, we can assume that any connection strings won't have any userinfo in them.

BTW: the specification includes a reference implementation, which can help guide on the order of splitting and what not: https://github.com/mongodb/specifications/blob/master/source/connection-string/connection-string-spec.rst#reference-implementation

For example, after we perform step 2, we can assume only "User information and host identifiers" part will be present for the moment, and after we perform step 3 we can assume only the "host identifiers" part will be present and skip step 4 altogether. In later PRs, we can incrementally add logic in for each of the steps to build up the functionality step by step.

Reply:
deleted